macSKK/KeyBinding.swift

```diff
@@ -167,7 +146,24 @@ struct KeyBinding: Identifiable {
         }
 
         func encode() -> [String: Any] {
-            return ["key": key.encode(), "modifierFlags": modifierFlags.rawValue, "displayString": displayString]
+            return [
+                "key": key.encode(),
+                "modifierFlags": modifierFlags.rawValue,
+                "optionalModifierFlags": optionalModifierFlags.rawValue,
+                "displayString": displayString,
+            ]
+        }
+
+        /// このキーバインド設定がキーイベントをこのキーバインドの入力として受理するかどうかを返す。
+        func accept(event: NSEvent) -> Bool {
+            if let character = event.charactersIgnoringModifiers?.lowercased().first, Key.characters.contains(character) {
+                if key != .character(character) {
+                    return false
+                }
+            } else if key != .code(event.keyCode) {
+                return false
+            }
```

Comment:
めちゃくちゃ微粒子レベルの指摘にはなりますが、`else if`でやっている比較のほうがなんか単純な比較だけなので、先にそっちをやったほうが、そっちでもう`return false`となったら` Key.characters.contains`をやらなくてよくなって超微妙に効率がよかったりますかね？

```suggestion
            if key != .code(event.keyCode) {
                return false
            } else if let character = event.charactersIgnoringModifiers?.lowercased().first, Key.characters.contains(character) {
                if key != .character(character) {
                    return false
                }
            }
```

Reply:
このaccept、最大で登録されているキーバインドの回数だけ呼ばれるのでそれも含めて富豪的な書き方になってますね。
ちゃんとやるならNSEventを引数にするのを止めて、ここでやってるちょっとした前処理を初期化で済ませた別の構造体に入れておくかなあとか思います。

```swift
struct CurrentInput {
  let key: Key
  let modifierFlags: NSEvent.ModifierFlags

  init(event: NSEvent) {
     if let character = event.charactersIgnoringModifiers?.lowercased().first, Key.characters.contains(character) {
       self.key = .character(character)
     }
     ...
  }
}
```

みたいな。

Reply:
なるほど、そうですね 👍
もともとは`NSEvent`から`KeyBinding.Input`をインスタンシーエションしていて、`KeyBinding`に用いる`Input`というデータ構造と入力の`modifierFlags`やキーコードなどを整理しておくデータ構造は別であったほうがいいと思ってました。
今回のPRの変更では、やや副作用かもしれませんが`KeyBinding.Input`が`NSEvent`からインスタンシーエションされるパスが消滅し`KeyBinding.Input`が真に`KeyBinding`用のデータ構造になったと思います。いったんこのPRの段階では今の状態でも十分とは思いますが、将来的には`CurrentInput`のようなデータ構造が`KeyBinding.Input`とは別に定義されるのはとてもいいと思いました 👍